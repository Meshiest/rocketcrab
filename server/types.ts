import { LobbyStatus } from "./enums";

export type RocketCrab = {
    lobbyList: Array<Lobby>;
};

export type Lobby = {
    status: LobbyStatus;
    playerList: Array<Player>;
    code: string;
};

export type Player = {
    name: string;
    socket: SocketIO.Socket;
};

export type Game = {
    newUrl: string;
};

export type JoinLobbyResponse = {
    code: string;
    name?: string;
};
