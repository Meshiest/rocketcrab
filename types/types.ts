import { PartyStatus, GameStatus } from "./enums";
import WebSocket from "ws";
import { Dispatch, SetStateAction } from "react";
import { BindingsChangeTarget } from "@geist-ui/react/dist/input/use-input";

export type RocketCrab = {
    partyList: Array<Party>;
};

export type Party = ClientParty & {
    nextPlayerId: number;
};

export type ClientParty = {
    status: PartyStatus;
    playerList: Array<Player>;
    code: string;
    uuid: string;
    me?: Player;
    selectedGameId: string;
    gameState: GameState;
    idealHostId: number;
};

export type Player = {
    id: number;
    name: string;
    socket?: SocketIO.Socket;
    isHost: boolean;
};

export type ClientGame = {
    id: string;
    name: string;
    author?: string;
    basedOn?: BasedOn;
    description?: string;
    displayUrlText: string;
    displayUrlHref: string;
    donationUrlText?: string;
    donationUrlHref?: string;
    guideId?: string;
    guide?: string;
    guideUrl?: string;
    pictures?: Array<string>;
    category?: Array<string>;
    players?: string;
    familyFriendly: boolean;
    minPlayers?: number;
    maxPlayers?: number;
    renameParams?: Record<string, string>;
};

export type BasedOn = {
    game: string;
    author?: string;
    link?: string;
};

export type ServerGame = ClientGame & {
    getJoinGameUrl: () => Promise<JoinGameURL>;
};

export type JoinGameURL = {
    playerURL: string;
    hostURL?: string;
    code?: string;
};

export type GameCategory = {
    id: string;
    name: string;
    color: string;
    backgroundColor: string;
};

export type ClientGameLibrary = {
    categories: Array<GameCategory>;
    gameList: Array<ClientGame>;
};

export type ServerGameLibrary = {
    categories: Array<GameCategory>;
    gameList: Array<ServerGame>;
};

export type GameState = {
    status: GameStatus;
    error?: string;
    joinGameURL?: JoinGameURL;
};

export type JoinPartyResponse = {
    code: string;
    lastPartyState: ClientParty;
    reconnecting: boolean;
};

export type PromiseWebSocket = WebSocket & {
    onOpen: () => Promise<void>;
    onMessage: () => Promise<string>;
    untilMessage: (msgChecker: (msg: string) => boolean) => Promise<string>;
};

export type MenuButton = {
    label: string;
    hostOnly: boolean;
    onClick: () => void;
};

export type LibraryState = {
    selectedCategory: string;
    setSelectedCategory: Dispatch<SetStateAction<string>>;
    search: string;
    setSearch: Dispatch<SetStateAction<string>>;
    searchBindings: SearchBindingsType;
};

type SearchBindingsType = {
    value: string;
    onChange: (event: BindingsChangeTarget) => void;
};
