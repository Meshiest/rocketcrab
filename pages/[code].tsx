import { useRouter } from "next/router";
import { useEffect, useState } from "react";
import socketIOClient from "socket.io-client";

import PageLayout from "../components/organisms/PageLayout";
import Lobby from "../components/organisms/Lobby";
import NameEntry from "../components/organisms/NameEntry";

const socket = socketIOClient();

export const Code = () => {
    const router = useRouter();
    const { code } = router.query;

    const [lobbyState, setLobbyState] = useState(initLobbyState());
    const { status } = lobbyState;

    // only ran with initial value due to the []
    useEffect(() => {
        socket.on("update", (newLobbyState) => {
            setLobbyState(newLobbyState);
        });

        return () => {
            socket.close();
            setLobbyState(initLobbyState());
        };
    }, []);

    useEffect(() => {
        // "code" will be undefined during Automatic Static Optimization
        if (code) socket.emit("join-lobby", { code });

        socket.on("invalid-lobby", () => router.push("/join?invalid=" + code));
    }, [code]);

    const onNameEntry = (name) => {
        socket.emit("name", name);
        // TODO: store in cookie
    };

    const { me } = lobbyState;

    const showLoading = status === "loading";
    const showNameEntry = !showLoading && !me.name;
    const showLobby = !showNameEntry && status.startsWith("lobby");

    return (
        <PageLayout path={code as string}>
            {showLoading && <div>Loading...</div>}
            {!showLoading && (
                <>
                    {showNameEntry && (
                        <NameEntry
                            onNameEntry={onNameEntry}
                            code={code}
                            socket={socket}
                        />
                    )}
                    {showLobby && (
                        <>
                            <Lobby playerList={lobbyState.playerList} />
                            <div>{JSON.stringify(lobbyState)}</div>
                        </>
                    )}
                </>
            )}
        </PageLayout>
    );
};

const initLobbyState = () => ({
    status: "loading",
    playerList: [],
    me: { name: undefined },
});

export default Code;
