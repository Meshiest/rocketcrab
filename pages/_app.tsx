import { ZeitProvider, CssBaseline } from "@zeit-ui/react";
import Body from "../components/templates/Body";
import "fontsource-inconsolata";
import { AppPropsType } from "next/dist/next-server/lib/utils";
import { initGA, logPageView } from "../utils/analytics";
import { useState, useEffect } from "react";
import Router from "next/router";

const MyApp = ({ Component, pageProps }: AppPropsType): JSX.Element => {
    const [, setLoading] = useState(false);
    useEffect(() => {
        initGA();
        logPageView();

        const loadingStart = () => setLoading(true);
        const loadingStop = () => {
            logPageView();
            setLoading(false);
        };

        Router.events.on("routeChangeStart", loadingStart);
        Router.events.on("routeChangeComplete", loadingStop);

        return () => {
            Router.events.off("routeChangeStart", loadingStart);
            Router.events.off("routeChangeComplete", loadingStop);
        };
    }, []);
    return (
        <ZeitProvider>
            <CssBaseline />
            <Body>
                <Component {...pageProps} />
            </Body>
        </ZeitProvider>
    );
};

export default MyApp;
