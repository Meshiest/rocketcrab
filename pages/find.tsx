import PrimaryButton from "../components/common/PrimaryButton";
import ButtonGroup from "../components/common/ButtonGroup";
import PageLayout from "../components/layout/PageLayout";
import React, { useEffect, useState } from "react";
import { postJson } from "../utils/utils";
import { useRouter } from "next/router";
import { io } from "socket.io-client";
import { SocketEvent } from "../types/enums";
import { ClientGameLibrary, FinderState } from "../types/types";
import { GetServerSideProps } from "next";
import { getClientGameLibrary } from "../config";
import PublicGame from "../components/find/PublicGame";
import { Spacer } from "@geist-ui/react";

const socket = io();
const CLIENT_GAME_LIBRARY = getClientGameLibrary();

export const Find = ({
    gameLibrary = { gameList: [], categories: [] },
}: FindProps): JSX.Element => {
    const router = useRouter();
    const [newLoading, setNewLoading] = useState(false);
    const [showReconnecting, setShowReconnecting] = useState(false);
    const [finderState, setFinderState] = useState<FinderState | undefined>();

    const onClickNew = async (e) => {
        e.preventDefault();
        setNewLoading(true);

        const { code } = await postJson("/api/new-public");

        router.push("/" + code);
    };

    useEffect(() => {
        socket.open();

        socket.emit(SocketEvent.FINDER_SUBSCRIBE);

        socket.on(SocketEvent.FINDER_UPDATE, (newFinderState: FinderState) => {
            setFinderState(newFinderState);
            setShowReconnecting(false);
        });

        socket.on(SocketEvent.DISCONNECT, (reason) => {
            // if the disconnection was initiated by the server
            if (reason === "io server disconnect") {
                // reconnect manually
                socket.connect();
            }
            // else the socket will automatically try to reconnect

            setShowReconnecting(true);
        });

        return () => {
            socket.close();
            setFinderState(undefined);
        };
    }, []);

    return (
        <PageLayout reconnecting={showReconnecting}>
            <div className="description">Public Parties</div>
            <div>
                {finderState?.publicPartyList?.map((party) => (
                    <PublicGame
                        key={party.code}
                        party={party}
                        gameLibrary={gameLibrary}
                    />
                ))}
                {!finderState?.publicPartyList?.length && (
                    <>
                        No public parties found. ðŸ˜ž You should make one! ðŸ¥°
                        <Spacer y={1.5} />
                    </>
                )}
            </div>
            <ButtonGroup>
                <PrimaryButton href="/" size="large">
                    Back
                </PrimaryButton>

                <PrimaryButton
                    onClick={onClickNew}
                    loading={newLoading}
                    size="large"
                >
                    Start Public Party
                </PrimaryButton>
            </ButtonGroup>
            <style jsx>{`
                .description {
                    text-align: center;
                    margin-bottom: 1em;
                }
            `}</style>
        </PageLayout>
    );
};

export const getServerSideProps: GetServerSideProps = async () => ({
    props: {
        gameLibrary: CLIENT_GAME_LIBRARY,
    },
});

type FindProps = {
    gameLibrary: ClientGameLibrary;
};

export default Find;
