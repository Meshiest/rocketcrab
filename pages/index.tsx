import Head from "next/head";
import Footer from "../components/Footer";
import MainTitle from "../components/MainTitle";
import PrimaryButton from "../components/PrimaryButton";
import ButtonGroup from "../components/ButtonGroup";

export const Home = (): JSX.Element => (
    <div className="container">
        <Head>
            <title>ðŸš€ðŸ¦€ rocketcrab</title>
        </Head>

        <main>
            <MainTitle />

            <div className="description">
                all the best mobile web party games
            </div>

            <ButtonGroup>
                <PrimaryButton
                    onClick={() => {
                        window.alert("With typescript and Jest");
                    }}
                >
                    Join Lobby
                </PrimaryButton>

                <PrimaryButton
                    onClick={() => {
                        window.alert("With typescript and Jest");
                    }}
                >
                    New Lobby
                </PrimaryButton>
            </ButtonGroup>
        </main>

        <Footer />

        <style jsx>{`
            .container {
                min-height: 100vh;
                padding: 0 0.5rem;
                display: flex;
                flex-direction: column;
                justify-content: center;
                align-items: center;
            }

            main {
                padding: 5rem 0;
                flex: 1;
                display: flex;
                flex-direction: column;
                justify-content: center;
                align-items: center;
            }

            a {
                color: inherit;
                text-decoration: none;
            }

            .description {
                text-align: center;
                margin-bottom: 2em;
            }
        `}</style>

        <style jsx global>{`
            @import url("https://fonts.googleapis.com/css2?family=Inconsolata:wght@400;600&display=swap");

            html,
            body {
                padding: 0;
                margin: 0;
            }

            * {
                box-sizing: border-box;
                letter-spacing: normal !important;
                font-family: "Inconsolata", monospace;
            }
        `}</style>
    </div>
);

export default Home;
