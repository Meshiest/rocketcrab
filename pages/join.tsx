import { useRouter } from "next/router";
import { Input, useInput } from "@zeit-ui/react";

import PrimaryButton from "../components/atoms/PrimaryButton";
import ButtonGroup from "../components/molecules/ButtonGroup";
import PageLayout from "../components/templates/PageLayout";

export const Join = (): JSX.Element => {
    const router = useRouter();
    const { state: code, bindings } = useInput("");

    const onEnter = (e) => {
        if (e.key !== "Enter") return;

        router.push("/[code]", "/" + code);
    };

    return (
        <PageLayout path={code}>
            <div className="description">Join Lobby</div>
            <div className="input-container">
                <Input
                    placeholder="ex. abcd"
                    size="large"
                    width="8rem"
                    clearable
                    maxLength={4}
                    autoFocus
                    onKeyDown={onEnter}
                    {...bindings}
                />
            </div>

            <ButtonGroup>
                <PrimaryButton href="/">Back</PrimaryButton>

                <PrimaryButton href="/[code]" as={"/" + code}>
                    Join
                </PrimaryButton>
            </ButtonGroup>

            <style jsx>{`
                .description {
                    text-align: center;
                    margin-bottom: 1em;
                }
                .input-container {
                    margin-bottom: 2em;
                    text-align: center;
                }
            `}</style>
        </PageLayout>
    );
};

export default Join;
