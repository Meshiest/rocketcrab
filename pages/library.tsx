import { GetServerSideProps } from "next";
import { useRouter } from "next/router";
import React, { useCallback } from "react";
import GameLibrary from "../components/molecules/GameLibrary";
import PageLayout from "../components/templates/PageLayout";
import { getClientGameLibrary } from "../config";
import { ClientGameLibrary } from "../types/types";

const CLIENT_GAME_LIBRARY = getClientGameLibrary();

export const Library = ({ gameLibrary }: LibraryProps): JSX.Element => {
    const router = useRouter();

    const onDone = useCallback(() => router.back(), [router]);

    const onViewGame = useCallback(
        (gameId) => {
            router.push("/game/[gameid]", "/game/" + gameId);
        },
        [router]
    );

    return (
        <PageLayout>
            <GameLibrary
                gameLibrary={gameLibrary}
                backToLabel="rocketcrab.com"
                onDone={onDone}
                setViewingGameId={onViewGame}
            />
        </PageLayout>
    );
};

type LibraryProps = {
    gameLibrary: ClientGameLibrary;
};

export const getServerSideProps: GetServerSideProps = async (): Promise<
    any
> => {
    return {
        props: {
            gameLibrary: CLIENT_GAME_LIBRARY,
        },
    };
};

export default Library;
