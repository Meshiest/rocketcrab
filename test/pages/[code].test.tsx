import React from "react";
import { render } from "../testUtils";
import { Code } from "../../pages/[code]";
import { ClientGameLibrary } from "../../types/types";

jest.mock("next/router", () => ({
    useRouter: () => ({ query: "ijkl" }),
}));

jest.mock("socket.io-client", () => ({
    __esModule: true,
    default: () =>
        (({
            open: jest.fn(),
            on: jest.fn(),
            close: jest.fn(),
        } as unknown) as SocketIO.Socket),
}));

describe("pages/[code].tsx", () => {
    it("matches snapshot", () => {
        const { asFragment } = render(
            <Code gameLibrary={{} as ClientGameLibrary} />,
            {}
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
