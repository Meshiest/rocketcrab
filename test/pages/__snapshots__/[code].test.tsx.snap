// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/[code].tsx matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-903643084 container"
  >
    <main
      class="jsx-903643084"
    >
      <div
        class="jsx-903643084 main-content"
      >
        <div
          class="jsx-3668335743 title"
          style="text-align: center;"
        >
          <h2
            class="jsx-3668335743"
          >
            ðŸš€ðŸ¦€
          </h2>
          <h2
            class="jsx-3668335743 lobby-url"
          >
            rocketcrab.com
          </h2>
        </div>
        <div
          class="jsx-3429833361 loading-container"
        >
          <span
            class="jsx-3429833361 loading"
          >
            <i
              class="jsx-3429833361"
            />
            <i
              class="jsx-3429833361"
            />
            <i
              class="jsx-3429833361"
            />
          </span>
        </div>
      </div>
    </main>
    <footer
      class="jsx-3128977650"
      style="margin-top: 1em;"
    >
      rocketcrab by 
      <a
        class="jsx-3128977650"
        href="https://www.tannerkrewson.com/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Tanner Krewson
      </a>
      <br
        class="jsx-3128977650"
      />
      <a
        class="jsx-3128977650"
        href="https://github.com/tannerkrewson/rocketcrab"
        rel="noopener noreferrer"
        target="_blank"
      >
        View on GitHub
      </a>
    </footer>
  </div>
</DocumentFragment>
`;
